baram/ec2_manager.py

```diff
@@ -1,105 +1,290 @@
 import boto3
+import traceback
+
+from baram.log_manager import LogManager
 
 
 class EC2Manager(object):
     def __init__(self):
         self.cli = boto3.client('ec2')
+        self.logger = LogManager.get_logger()
 
     def list_security_groups(self):
-        '''
+        """
         Describes the specified security groups or all of your security groups.
 
         :return: SecurityGroups
-        '''
+        """
         return self.cli.describe_security_groups()['SecurityGroups']
 
+    def list_redundant_security_group_ids(self, redundant_domain_ids: list = None):
+        """
+        Describe useless and deletable security group ids.
+        (including disused SageMaker domain related things)
+
+        :param redundant_domain_ids:
+        :return: GroupIds
+        """
+        # 1. Get security groups not related to anything in ec2
+        vpc_sg_eni_subnets = self.list_vpc_sg_eni_subnets()
+        security_group_valid_ids = set([pair['sg_id'] for pair in vpc_sg_eni_subnets])
+
+        security_groups = self.list_security_groups()
+        security_group_ids_total = set([sg['GroupId'] for sg in security_groups])
+
+        result = security_group_ids_total - security_group_valid_ids
+
+        # 2. Get security groups related to disused domains; SageMaker (NFS related)
+        if redundant_domain_ids is not None:
+            nfs_sgs = [sg['GroupId'] for sg in security_groups
+                       if 'NFS' in sg['Description']
+                       and sum([domain_id in sg['Description'] for domain_id in redundant_domain_ids]) == 0]
+            result.union(nfs_sgs)
```

Comment:
two issues.

1. too SM specific.
2. It's better to filter codition at line 30 and 33 if it's possible.

Reply:
I added those lines because making additional method like `get_sg_id_for_unused_sms` looked too minor. After reading this comment, however, It feels like it is not that minor thing. Is this thought reasonable?

Reply:
I think you can make argument as 'description_filter' and inject (such as) 'NFS'.

It may function same way but it's an generalization of a method.